content/en/docs/concepts/configuration/manage-resources-containers.md

```diff
@@ -34,16 +34,20 @@ For example, if you set a `memory` request of 256 MiB for a container, and that
 a Pod scheduled to a Node with 8GiB of memory and no other Pods, then the container can try to use
 more RAM.
 
-If you set a `memory` limit of 4GiB for that container, the kubelet (and
-{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}) enforce the limit.
-The runtime prevents the container from using more than the configured resource limit. For example:
-when a process in the container tries to consume more than the allowed amount of memory,
-the system kernel terminates the process that attempted the allocation, with an out of memory
-(OOM) error.
+Both `CPU` and `memory` limits are applied by the kubelet (and
+{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}),
+and are ultimately enforced by the kernel. On Linux nodes, limits are enforced
+by {{< glossary_tooltip text="cgroups" term_id="cgroup" >}}). However, `CPU` and `memory` limits
+are enforced differently.
 
-Limits can be implemented either reactively (the system intervenes once it sees a violation)
-or by enforcement (the system prevents the container from ever exceeding the limit). Different
-runtimes can have different ways to implement the same restrictions.
+`CPU` limits are enforced by CPU throttling. When a container approaches
+its `CPU` limit, the kernel will restrict access to the CPU corresponding to the
+container's limit.
+
+`memory` limits are enforced by the kernel with OOM (out of memory) kills. When
+a pod uses more than its `memory` limit, the kernel may terminate it. However,
+if there is available memory on the node, the kernel may choose not to terminate
+the offending pod.
```

Comment:
1. What happens to memory allocation attempts -  `sbrk()` and friends - that would assign memory above the limit, if they were to succeed? Ideally, let's be clear about where the enforcement does and doesn't happen.

1. The container runtime or kernel might apply a higher limit than you request. For example, if you specify a limit of 2097159 bytes (a shade over 2MiB), the kernel is likely to round things up to a whole page size. That's if your container runtime even lets you request a limit that small!

Reply:
I don't agree to this change related to memory limits. It is not accurate and it will lead to more confusion, because memory limit enforcement is a pretty complex topic.

Reply:
Is it because the kubelet or container runtime uses a [cgroup driver](https://kubernetes.io/docs/setup/production-environment/container-runtimes/#cgroup-drivers) to interact with the OS’s underlying cgroup system to enforce a limit, hence, any limit created by a different cgroup does not matter?

Reply:
@network-charles Did you read this? https://kubernetes.io/blog/2023/05/05/qos-memory-resources/

Reply:
I just did. It's similar to my last comment but with an added info on the “memory.high” feature supported by cgroupv2. It's clear now.

We may need to clarify that the kernel uses the “memory. high” feature supported on cgroupv2 to throttle the container. Also, the statement that it applies to a pod is wrong, according to the blog, it only applies at the container cgroups level.

Reply:
So ... the point is ... a Pod may and may not get killed when it is using more memory than its limit.
Can we rephrase this to be more accurate?

Reply:
I added a note about memoryQOS feature, it's not something we should document heavily because the effort is stalled in SIG node due to issues

Reply:
> What happens to memory allocation attempts - sbrk() and friends - that would assign memory above the limit, if they were to succeed? Ideally, let's be clear about where the enforcement does and doesn't happen.

it's unfortunately not so easy. When the kernel enforces OOM kills, it usually is in response to a memory allocation. so sometimes, sbrk and friends will trigger an oom kill of the process, and other times it will succeed.

> The container runtime or kernel might apply a higher limit than you request. For example, if you specify a limit of 2097159 bytes (a shade over 2MiB), the kernel is likely to round things up to a whole page size. That's if your container runtime even lets you request a limit that small!

I don't know if we need to get to this level of detail here, but I'm open to suggestions